common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,893 @@
+    // The offset for the last chunk tracked in the index file for this shuffle partition
+    private long lastChunkOffset;
+    private int lastMergedMapId = -1;
+
+    // Bitmap tracking which mapper's blocks are in the current shuffle chunk
+    private RoaringBitmap chunkTracker;
+    ByteBuf trackerBuf = null;
+
+    AppShufflePartitionInfo(
+        AppShufflePartitionId partitionId,
+        File dataFile,
+        File indexFile,
+        File metaFile) throws IOException {
+      this.partitionId = Preconditions.checkNotNull(partitionId, "partition id");
+      dataFile.createNewFile();
+      this.dataFile = dataFile;
+      this.channel = new FileOutputStream(dataFile, true).getChannel();
+      indexFile.createNewFile();
+      FileOutputStream fos = new FileOutputStream(indexFile, true);
+      indexChannel = fos.getChannel();
+      this.indexWriteStream = new DataOutputStream(new BufferedOutputStream(fos));
+      metaFile.createNewFile();
+      metaChannel = new FileOutputStream(metaFile, true).getChannel();
+      this.currentMapId = -1;
+      // Writing 0 offset so that we can reuse ShuffleIndexInformation.getIndex()
+      updateChunkInfo(0L, -1);
+      this.position = 0;
+      this.encounteredFailure = false;
+      this.mapTracker = new RoaringBitmap();
+      this.chunkTracker = new RoaringBitmap();
+    }
+
+    public long getPosition() {
+      return position;
+    }
+
+    public void setPosition(long position) {
+      logger.trace("{} shuffleId {} reduceId {} current pos {} update pos {}", partitionId.appId,
+        partitionId.shuffleId, partitionId.reduceId, this.position, position);
+      this.position = position;
+    }
+
+    boolean isEncounteredFailure() {
+      return encounteredFailure;
+    }
+
+    void setEncounteredFailure(boolean encounteredFailure) {
+      this.encounteredFailure = encounteredFailure;
+    }
+
+    int getCurrentMapId() {
+      return currentMapId;
+    }
+
+    void setCurrentMapId(int mapId) {
+      logger.trace("{} shuffleId {} reduceId {} updated mapId {} current mapId {}",
+        partitionId.appId, partitionId.shuffleId, partitionId.reduceId, currentMapId, mapId);
+      this.currentMapId = mapId;
+    }
+
+    long getLastChunkOffset() {
+      return lastChunkOffset;
+    }
+
+    void blockMerged(int mapId) {
+      logger.debug("{} shuffleId {} reduceId {} updated merging mapId {}", partitionId.appId,
+        partitionId.shuffleId, partitionId.reduceId, mapId);
+      mapTracker.add(mapId);
+      chunkTracker.add(mapId);
+      lastMergedMapId = mapId;
+    }
+
+    void resetChunkTracker() {
+      chunkTracker.clear();
+    }
+
+    /**
+     * Appends the chunk offset to the index file and adds the mapId to the chunk tracker.
+     *
+     * @param chunkOffset the offset of the chunk in the data file.
+     * @param mapId the mapId to be added to chunk tracker.
+     */
+    void updateChunkInfo(long chunkOffset, int mapId) throws IOException {
```

Comment:
nit: rename `mapId` to `mapIndex` and the param doc.

Reply:
Done